spec/app/ics-029-fee-payment/README.md

```diff
@@ -0,0 +1,223 @@
+the packet.
+
+Given this, the flow would be:
+
+1. User/module submits a send packet on the `source` chain, along with some tokens and fee information on how to distribute them. The fee tokens are all escrowed by the fee module.
+2. Relayer submits `OnReceivePacket` on the `destination` chain. Along with this message, the *forward relayer* will submit a `payToOnSource` address where payment should be sent.
+3. Destination application includes this `payToOnSource` in the acknowledgement (there are multiple approaches to discuss below)
+4. Relayer submits `OnAcknowledgement` which provides the *return relayer* address on the source chain, along with the `payToOnSource` address
+5. Source application can distribute the tokens escrowed in (1) to both the *forward* and the *return* relayers.
+
+Alternate flow:
+
+1. User/module submits a send packet on the `source` chain, along with some tokens and fee information on how to distribute them
+2. Relayer submits `OnTimeout` which provides its address on the source chain
+3. Source application can distribute the tokens escrowed in (1) to this relayer, and potentially return remainder tokens to the original packet sender.
+
+### Fee details
+
+For an example implementation in the Cosmos SDK, we consider 3 potential fee payments, which may be defined. Each one may be
+paid out in a different token. Imagine a connection between IrisNet and the Cosmos Hub. They may define:
+
+- ReceiveFee: 0.003 channel-7/ATOM vouchers (ATOMs already on IrisNet via ICS20)
+- AckFee: 0.001 IRIS
+- TimeoutFee: 0.002 IRIS
+
+Ideally the fees can easily be redeemed in native tokens on both sides, but relayers may select others. In this example, the relayer collects a fair bit of IRIS, covering it's costs there and more. It also collects channel-7/ATOM vouchers from many packets. After relaying a few thousand packets, the account on the Cosmos Hub is running low, so the relayer will send those channel-7/ATOM vouchers back over channel-7 to it's account on the Hub to replenish the supply there. 
+
+The sender chain will escrow 0.003 channel-7/ATOM and 0.002 IRIS. In the case that a forward relayer submits the `MsgRecvPacket` and a reverse relayer submits the `MsgAckPacket`, the forward relayer is reqarded 0.003 channel-7/ATOM and the reverse relayer is rewarded 0.001 IRIS. In the case where the packet times out, the timeout relayer receives 0.002 IRIS and 0.003 channel-7/ATOM is refunded to the original fee payer.
+
+The logic involved in collecting fees from users and then paying it out to the relevant relayers is encapsulated by a separate fee module and may vary between implementations. However, all fee modules must implement a uniform interface such that the ICS-4 handlers can correctly pay out fees to the right relayers, and so that relayers themselves can easily determine the fees they can expect for relaying a packet.
```

Comment:
Why do we want to do this from the ICS-4 handlers? Is it possible to split this out into middleware? I don't really like including fee logic in core IBC since it has nothing to do with correctness of transport, authentication, or ordering semantics.

Reply:
I think the main question is do we enable this feature on a per-app basis, or on a per-chain basis.

I think there's consensus to enable it per-chain: https://github.com/cosmos/ibc/pull/581#discussion_r644758984

I'm agnostic on whether we implement this uniform behaviour directly into ICS-4 handlers, or create some middleware that wraps all application callbacks.

cc: @ethanfrey

Reply:
I agree with pulling the actual fee handling logic out of ICS-4, but rather define an interface for it.

We could just have a middleware that fulfills the application interface and then calls into another application. Or we could have something like an `AnteHandler` interface, which is called before dispatching to the application.

There have been requests to keep the application interface stable (and not expose the `payOnSource` info there, just the raw acknowledgement bytes). Thus, we couldn't just wrap one application with a fee handler decorator in the same interface. If we do pass all info down, this would be a valid approach IMO. 

If we let app "decorate" it's handlers to allow fee-processing, incentivization may be missing on many cases, which could be bad or good. (I have heard mention of it being opt-in)

I am more in favor of defining some "AnteHandler" interface in ICS4/ICS26 that gets the full metadata on packet and acknoweldgement and relayer/signer. The fee handler must fit into such an interface (but functionality be defined in it's own application-level spec). We could later add other "Ante Handlers" without requiring changes in the core handlers or logic.

Reply:
"AnteHandler" allows us to enable this on a per-chain basis, without requiring more than an optional callback in ICS-4.